core/src/main/java/com/linecorp/armeria/client/HAProxyHandler.java

```diff
@@ -0,0 +1,113 @@
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static io.netty.handler.codec.haproxy.HAProxyCommand.PROXY;
+import static io.netty.handler.codec.haproxy.HAProxyProtocolVersion.V2;
+import static io.netty.handler.codec.haproxy.HAProxyProxiedProtocol.TCP4;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.proxy.HAProxyConfig;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.haproxy.HAProxyMessage;
+import io.netty.handler.codec.haproxy.HAProxyMessageEncoder;
+import io.netty.handler.proxy.ProxyConnectException;
+
+class HAProxyHandler extends ChannelOutboundHandlerAdapter {
+    private static final Logger logger = LoggerFactory.getLogger(HAProxyHandler.class);
+
+    private final HAProxyConfig haProxyConfig;
+
+    HAProxyHandler(HAProxyConfig haProxyConfig) {
+        this.haProxyConfig = haProxyConfig;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline().addBefore(ctx.name(), null, HAProxyMessageEncoder.INSTANCE);
+        super.handlerAdded(ctx);
+    }
+
+    @Override
+    public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress,
+                        SocketAddress localAddress, ChannelPromise promise) throws Exception {
+        promise.addListener(f -> {
+            if (!f.isSuccess()) {
+                return;
+            }
+            try {
+                ctx.write(createMessage(haProxyConfig, ctx.channel())).addListener(f0 -> {
+                    if (f0.isSuccess()) {
+                        ctx.pipeline().remove(HAProxyMessageEncoder.INSTANCE);
+                    } else {
+                        ctx.fireExceptionCaught(new ProxyConnectException(f0.cause()));
+                        ctx.close();
+                    }
+                });
+            } catch (Exception e) {
+                ctx.channel().eventLoop().execute(
+                        () -> ctx.pipeline().fireUserEventTriggered(new ProxyConnectException(e)));
+                ctx.close();
```

Comment:
Shouldn't we call `ctx.close()` after calling `fireUserEventTriggered()`?

Reply:
🦅  👁️.

I also found another place doing this:
https://github.com/line/armeria/blob/4d417ea1fadb416c6ba2be8e46c34acd5249dfb8/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java#L372-L375

The behavior would be only slightly less informative exception messages, so I don't think going through the code and correcting all instances of this is critical though.